File: src/com/easy/RemoveDuplicatesFromSortedArray_26.java
/**
 * 
 */
package com.easy;

/**
 * @author Chris Lin
 * @version 1.0
 */
public class RemoveDuplicatesFromSortedArray_26 {
    // Use index count
    // Time Complexity: O(n)
    // Space Complexity:O(1)
    // Runtime: 0 ms, faster than 100.00%
    // Memory Usage: 38.8 MB, less than 99.99%
    public int removeDuplicates(int[] nums) {
        if (nums.length == 0) {
            return 0;
        }
        int mIndex = 0;
        for (int i = 1; i < nums.length; i++) {
            if (nums[i] != nums[i - 1]) {
                nums[++mIndex] = nums[i];
            }
        }

        return (mIndex + 1);
    }
}

File: Golang_src/easy/removeDuplicatesFromSortedArray_26.go
package easy

// Use two pointers:
// Time Complexity: O(n)
// Space Complexity:O(1)
// Runtime: 8 ms, faster than 88.17%
// Memory Usage: 4.6 MB, less than 100.00%
func removeDuplicates(nums []int) int {
	if nums == nil || len(nums) == 0 {
		return 0
	}

	result := 1
	idx := 0
	cur := nums[0]

	// O(n)
	for i := 1; i < len(nums); i++ {
		if nums[i] != cur {
			cur = nums[i]
			result++
			idx++
			nums[idx], nums[i] = nums[i], nums[idx]
		}
	}

	return result
}
